lint/golangci-lint.bzl

```diff
@@ -0,0 +1,146 @@
+    """
+
+    # golangci-lint calls out to Go, so we need the go context
+    go = go_context(ctx)
+    inputs = srcs + [config] + go.sdk_files
+    args = ctx.actions.args()
+    args.add_all(srcs)
+
+    command = """#!/usr/bin/env bash
+        export GOROOT=$(cd "$(dirname {go_tool})/.."; pwd)
+        export GOPATH=$GOROOT
+        export GOCACHE="$(mktemp -d)"
+        export PATH="$GOPATH/bin:$PATH"
+        GOLANGCI_LINT_CACHE=$(pwd)/.cache {golangci_lint} run --config={config} $@"""
+
+    if use_exit_code:
+        command += " && touch {report}"
+    else:
+        command += " 2>{report} || true"
+
+    ctx.actions.run_shell(
+        inputs = inputs,
+        outputs = [report],
+        command = command.format(
+            golangci_lint = executable.path,
+            report = report.path,
+            config = config.path,
+            go_tool = ctx.toolchains["@io_bazel_rules_go//go:toolchain"].sdk.go.path,
+        ),
+        env = go.env,
+        arguments = [args],
+        mnemonic = _MNEMONIC,
+        tools = [executable],
+    )
+
+# buildifier: disable=function-docstring
+def _golangci_lint_aspect_impl(target, ctx):
+    if ctx.rule.kind not in ["go_binary", "go_library", "go_test"]:
+        return []
+
+    report, info = report_file(_MNEMONIC, target, ctx)
+    golangci_lint_action(ctx, ctx.executable._golangci_lint, ctx.rule.files.srcs, ctx.file._config_file, report, ctx.attr.fail_on_violation)
+    return [info]
+
+def golangci_lint_aspect(binary, config):
+    """A factory function to create a linter aspect.
+
+    Attrs:
+        binary: a golangci-lint executable.
+        config: the .golangci.yaml file
+    """
+    return aspect(
+        implementation = _golangci_lint_aspect_impl,
+        attrs = {
+            "fail_on_violation": attr.bool(),
+            "_golangci_lint": attr.label(
+                default = binary,
+                executable = True,
+                cfg = "exec",
+            ),
+            "_config_file": attr.label(
+                default = config,
+                allow_single_file = True,
+            ),
+        },
+        toolchains = ["@io_bazel_rules_go//go:toolchain"],
+    )
+
+def golangci_lint_binary(name):
+    """Wrapper around native_binary to select the correct golangci-lint executable for the execution platform."""
+    native_binary(
+        name = name,
+        src = select(
+            {
+                "@bazel_tools//src/conditions:linux_x86_64": "@golangci_lint_linux_x86_64//:golangci-lint",
+                "@bazel_tools//src/conditions:linux_aarch64": "@golangci_lint_linux_aarch64//:golangci-lint",
+                "@bazel_tools//src/conditions:darwin_x86_64": "@golangci_lint_macos_x86_64//:golangci-lint",
+                "@bazel_tools//src/conditions:darwin_arm64": "@golangci_lint_macos_aarch64//:golangci-lint",
+            },
+            no_match_error = "golangci-lint hasn't been fetched for your platform",
```

Comment:
could you also say what the user ought to do about it? (file a bug, etc)

Reply:
Now that I am moving it to the example folder I am gonna get rid of it. It doesn't add much value, and the solution isn't always the same.